scipp_widgets/widgets.py

```diff
@@ -0,0 +1,146 @@
+
+    if (output_widget):
+        output_widget.clear_output()
+        with output_widget:
+            display(Javascript(output))
+    else:
+        display(Javascript(output))
+
+
+class HideCodeWidget(widgets.Box):
+    """
+    Button which toggles the visibilty of code block.
+    """
+    def __init__(self, state):
+        super().__init__()
+        self.output_widget = widgets.Output()
+        self.button = widgets.ToggleButton(not state, description='Py')
+        self.button.observe(self._on_button_toggle, "value")
+        self.button.layout.flex = '0 1 40px'
+        self.button.layout.align_self = 'flex-end'
+
+        self.children = [self.button, self.output_widget]
+        self.layout.align_self = 'flex-end'
+        toggle_code(not state)
+
+    def _on_button_toggle(self, value):
+        toggle_code(value.new, self.output_widget)
+
+
+class DisplayWidget(widgets.Box):
+    """
+    Provides a simple graphical wrapper around a given callable.
+    """
+    def __init__(self, callable, name: str, input_specs, hide_code=False):
+        """
+        Parameters:
+        callable (Callable): The function to call
+        name (str): name of widget to display
+        input_specs (list): class containing input data
+        """
+        super().__init__()
+        self.layout.flex_flow = 'column'
+        self.callable = callable
+        self.input_specs = input_specs
+        self.input_widgets = []
+        self._setup_input_widgets(input_specs)
+
+        self.button = widgets.Button(description=name)
+        self.button.on_click(self._on_button_clicked)
+        self.button_widgets = [self.button]
+        if (hide_code):
+            self.button_widgets += (HideCodeWidget(True), )
+
+        self.output_area = widgets.Output()
+        self.output_widgets = widgets.VBox([self.output_area])
+
+        self.row_widgets = widgets.HBox(self.input_widgets +
+                                        self.button_widgets)
+        self.row_widgets.layout.flex_flow = 'row wrap'
+
+        self.children = [self.row_widgets, self.output_widgets]
+
+    def _setup_input_widgets(self, inputs):
+        """
+        Creates a user-input widget for each item in inputs
+        """
+        for spec in inputs:
+            self.input_widgets.append(spec.create_input_widget())
+
+    def _retrive_kwargs(self):
+        kwargs = {
+            spec.name: spec.validate(widget.value)
+            for spec, widget in zip(self.input_specs, self.input_widgets)
+        }
+        return kwargs
+
+    def _on_button_clicked(self, button):
+        self.output_area.clear_output()
+        with self.output_area:
+            try:
+                kwargs = self._retrive_kwargs()
+            except ValueError as e:
+                print(f'Invalid inputs: {e}')
+                return
+
+            self._process(kwargs)
+
+    def _process(self, kwargs):
+        display(self.callable(**kwargs))
+
+
+class ProcessWidget(DisplayWidget):
```

Comment:
Feels odd to inherit from a *display* class here. Would it make sense to move the common functionality to something like `class WidgetBase`?

Reply:
Yes probably would in practice this will mean that DisplayWidget will just inherit the base widget and change little to nothing. But think it is worth it for the gain in code readability.